polar-core/src/vm.rs

```diff
@@ -905,57 +905,31 @@ impl PolarVirtualMachine {
         );
 
         match (&left.value(), &right.value()) {
-            (Value::List(left), Value::List(right)) => {
-                self.unify_lists(left, right, |(left, right)| Goal::Isa {
-                    left: left.clone(),
-                    right: right.clone(),
-                })?;
-            }
-
-            (Value::Dictionary(left), Value::Pattern(Pattern::Dictionary(right))) => {
-                // Check that the left is more specific than the right.
-                let left_fields: HashSet<&Symbol> = left.fields.keys().collect();
-                let right_fields: HashSet<&Symbol> = right.fields.keys().collect();
-                if !right_fields.is_subset(&left_fields) {
-                    return self.push_goal(Goal::Backtrack);
-                }
-
-                // For each field on the right, isa its value against the corresponding value on
-                // the left.
-                for (k, v) in right.fields.iter() {
-                    let left = left
-                        .fields
-                        .get(&k)
-                        .expect("left fields should be a superset of right fields")
-                        .clone();
+            (Value::Variable(symbol), _) => {
+                if let Some(value) = self.value(&symbol).cloned() {
                     self.push_goal(Goal::Isa {
-                        left,
-                        right: v.clone(),
-                    })?
+                        left: value,
+                        right: right.clone(),
+                    })?;
+                } else {
+                    self.push_goal(Goal::Unify {
+                        left: left.clone(),
+                        right: right.clone(),
+                    })?;
                 }
             }
 
-            (Value::InstanceLiteral(_), _) => {
-                panic!("How did an instance literal get here???");
-            }
-
-            (Value::ExternalInstance(_), Value::Pattern(Pattern::Dictionary(right))) => {
```

Comment:
This case was generalised to `(_, Value::Pattern(...))`, since we want it to do the same for _any_ type (except dictionaries which are handled first)